src/AggregateFunctions/AggregateFunctionTimeWeighted.cpp

```diff
@@ -0,0 +1,136 @@
+
+        return nested_arguments;
+    }
+
+    /// Decimal128 and Decimal256 aren't supported
+    #define AT_SWITCH(LINE) \
+        switch (which.idx) \
+        { \
+            LINE(Int8); LINE(Int16); LINE(Int32); LINE(Int64); LINE(Int128); LINE(Int256); \
+            LINE(UInt8); LINE(UInt16); LINE(UInt32); LINE(UInt64); LINE(UInt128); LINE(UInt256); \
+            LINE(Decimal32); LINE(Decimal64); \
+            LINE(Float32); LINE(Float64); \
+            default: return nullptr; \
+        }
+
+    // Not using helper functions because there are no templates for binary decimal/numeric function.
+    template <class... TArgs>
+    IAggregateFunction * create(const IDataType & first_type, const IDataType & second_type, TArgs && ... args) const
+    {
+        const WhichDataType which(first_type);
+
+    #define LINE(Type) \
+        case TypeIndex::Type:       return create<Type, TArgs...>(second_type, std::forward<TArgs>(args)...)
+        AT_SWITCH(LINE)
+    #undef LINE
+    }
+    template <class First, class ... TArgs>
+    IAggregateFunction * create(const IDataType & second_type, TArgs && ... args) const
+    {
+        const WhichDataType which(second_type);
+
+        switch (which.idx)
+        {
+            case TypeIndex::Date:    return new AggregateFunctionTimeWeighted<First, DataTypeDate::FieldType>(std::forward<TArgs>(args)...);
+            case TypeIndex::Date32:   return new AggregateFunctionTimeWeighted<First, DataTypeDate32::FieldType>(std::forward<TArgs>(args)...);
+            case TypeIndex::DateTime:   return new AggregateFunctionTimeWeighted<First, DataTypeDateTime::FieldType>(std::forward<TArgs>(args)...);
+            case TypeIndex::DateTime64: return new AggregateFunctionTimeWeighted<First, DataTypeDateTime64::FieldType>(std::forward<TArgs>(args)...);
+            default: return nullptr;
+        }
+    }
+
+    AggregateFunctionPtr transformAggregateFunction(
+        const AggregateFunctionPtr & nested_function,
+        const AggregateFunctionProperties &,
+        const DataTypes & arguments,
+        const Array & params) const override
+    {
+        if (arguments.size() != 3 && arguments.size() != 2)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Function {} should have two or three arguments",this->getName());
+
+        const auto data_type = static_cast<const DataTypePtr>(arguments[0]);
+        const auto data_type_time_weight = static_cast<const DataTypePtr>(arguments[1]);
+        const WhichDataType dt(data_type), t_dt(data_type_time_weight);
+
+        if ((dt.isInt() || dt.isUInt() || dt.isFloat() || dt.isDecimal()) && (t_dt.isDateOrDate32() || t_dt.isDateTime()|| t_dt.isDateTime64()))
+        else    
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Types {} and {} are non-conforming as arguments for aggregate function {}", data_type->getName(), data_type_time_weight->getName(), this->getName());
```

Comment:
`value type` and `time type` should be checked separately , and throw different error messages

Reply:
We don't need to check the first parameter, let the nested function check it.

just care for second and third time parameters